db/install.xml

```diff
@@ -49,5 +50,20 @@
         <INDEX NAME="reporttype_idx" UNIQUE="false" FIELDS="reporttype"/>
       </INDEXES>
     </TABLE>
+    <TABLE NAME="tool_objectfs_object_tags" COMMENT="Object tag key value pairs">
+      <FIELDS>
+        <FIELD NAME="id" TYPE="int" LENGTH="10" NOTNULL="true" SEQUENCE="true"/>
+        <FIELD NAME="contenthash" TYPE="char" LENGTH="40" NOTNULL="true" SEQUENCE="false"/>
+        <FIELD NAME="tagkey" TYPE="char" LENGTH="32" NOTNULL="true" SEQUENCE="false"/>
+        <FIELD NAME="tagvalue" TYPE="char" LENGTH="128" NOTNULL="true" SEQUENCE="false"/>
+        <FIELD NAME="timemodified" TYPE="int" LENGTH="10" NOTNULL="true" SEQUENCE="false"/>
+      </FIELDS>
+      <KEYS>
+        <KEY NAME="primary" TYPE="primary" FIELDS="id"/>
+      </KEYS>
+      <INDEXES>
+        <INDEX NAME="objecttagkey_idx" UNIQUE="true" FIELDS="contenthash, tagkey"/>
+      </INDEXES>
```

Comment:
I went back and forth on this myself for a lot.

Contenthash is so much simpler in terms of DB operations (less lookups, etc...) and code, because we don't have to lookup the object id (since almost all the existing objectfs apis use contenthash, not object id)

But you are right, it means we have a long string instead of an id, hence it takes up more storage.

I lent towards simplicity, but i'll look at it again and see if it can be implemented nicely without too many extra db operations.

Reply:
Realistically this isn't too much of a concern at the current time with only 2 classes of tags, so I'm happy to go with whatever you think is best.

Reply:
Had another look at it and really we only need the objectid when storing the tags, everything else can still be based off hash.

There were two options though to get the object id here:

1. Drill the objectid down from here https://github.com/catalyst/moodle-tool_objectfs/blob/d0f301b9bc4e2682119c81a434b236f2c860c655/classes/local/object_manipulator/pusher.php#L72 all the way to `push_object_tags`, which would save 1 db query but requires quite a bit of api changes just for this
2. Re-query it when saving tags just using `$DB->get_field`

I went with option 2 since its simpler code-wise and really 1 db query is not bad considering most objects will have that operation run once in its entire lifetime.

Lmk if you have different thoughts about this though

Reply:
I agree with option 2 if this is the only feature that would benefit from 1.